File: languages/go/builder/builder.go
package builder

import (
	"context"
	"errors"

	"github.com/bloock/bloock-sdk-go/v2/client/entity"
	"github.com/bloock/bloock-sdk-go/v2/internal/bridge"
	"github.com/bloock/bloock-sdk-go/v2/internal/bridge/proto"
	"github.com/bloock/bloock-sdk-go/v2/internal/config"
)

type RecordBuilder struct {
	payload     interface{}
	payloadType proto.RecordTypes
	signer      *proto.Signer
	encrypter   *proto.Encrypter
	decrypter   *proto.Decrypter
}

func (b RecordBuilder) WithSigner(signer entity.Signer) RecordBuilder {
	b.signer = signer.ToProto()
	return b
}

func (b RecordBuilder) WithEncrypter(encrypter entity.Encrypter) RecordBuilder {
	b.encrypter = encrypter.ToProto()
	return b
}

func (b RecordBuilder) WithDecrypter(decrypter entity.Decrypter) RecordBuilder {
	b.decrypter = decrypter.ToProto()
	return b
}

func (b RecordBuilder) Build() (entity.Record, error) {
	bridgeClient := bridge.NewBloockBridge()

	var res *proto.RecordBuilderResponse
	var err error

	switch b.payloadType {
	case proto.RecordTypes_STRING:
		res, err = bridgeClient.Record().BuildRecordFromString(context.Background(), &proto.RecordBuilderFromStringRequest{
			ConfigData: config.NewConfigData(),
			Payload:    b.payload.(string),
			Signer:     b.signer,
			Encrypter:  b.encrypter,
			Decrypter:  b.decrypter,
		})
	case proto.RecordTypes_BYTES:
		res, err = bridgeClient.Record().BuildRecordFromBytes(context.Background(), &proto.RecordBuilderFromBytesRequest{
			ConfigData: config.NewConfigData(),
			Payload:    b.payload.([]byte),
			Signer:     b.signer,
			Encrypter:  b.encrypter,
			Decrypter:  b.decrypter,
		})
	case proto.RecordTypes_FILE:
		res, err = bridgeClient.Record().BuildRecordFromFile(context.Background(), &proto.RecordBuilderFromFileRequest{
			ConfigData: config.NewConfigData(),
			Payload:    b.payload.([]byte),
			Signer:     b.signer,
			Encrypter:  b.encrypter,
			Decrypter:  b.decrypter,
		})
	case proto.RecordTypes_JSON:
		res, err = bridgeClient.Record().BuildRecordFromJson(context.Background(), &proto.RecordBuilderFromJSONRequest{
			ConfigData: config.NewConfigData(),
			Payload:    b.payload.(string),
			Signer:     b.signer,
			Encrypter:  b.encrypter,
			Decrypter:  b.decrypter,
		})
	case proto.RecordTypes_HEX:
		res, err = bridgeClient.Record().BuildRecordFromHex(context.Background(), &proto.RecordBuilderFromHexRequest{
			ConfigData: config.NewConfigData(),
			Payload:    b.payload.(string),
			Signer:     b.signer,
			Encrypter:  b.encrypter,
			Decrypter:  b.decrypter,
		})
	case proto.RecordTypes_RECORD:
		res, err = bridgeClient.Record().BuildRecordFromRecord(context.Background(), &proto.RecordBuilderFromRecordRequest{
			ConfigData: config.NewConfigData(),
			Payload:    b.payload.(*proto.Record),
			Signer:     b.signer,
			Encrypter:  b.encrypter,
			Decrypter:  b.decrypter,
		})

	case proto.RecordTypes_LOADER:
		res, err = bridgeClient.Record().BuildRecordFromLoader(context.Background(), &proto.RecordBuilderFromLoaderRequest{
			ConfigData: config.NewConfigData(),
			Loader:     b.payload.(*proto.Loader),
			Signer:     b.signer,
			Encrypter:  b.encrypter,
			Decrypter:  b.decrypter,
		})
	}

	if err != nil {
		return entity.Record{}, err
	}

	if res.Error != nil {
		return entity.Record{}, errors.New(res.Error.Message)
	}

	return entity.NewRecordFromProto(res.Record), nil
}

func NewRecordBuilderFromRecord(record entity.Record) RecordBuilder {
	return RecordBuilder{
		payload:     record.ToProto(),
		payloadType: proto.RecordTypes_RECORD,
	}
}

func NewRecordBuilderFromLoader(loader entity.Loader) RecordBuilder {
	return RecordBuilder{
		payload:     loader.ToProto(),
		payloadType: proto.RecordTypes_LOADER,
	}
}

func NewRecordBuilderFromString(str string) RecordBuilder {
	return RecordBuilder{
		payload:     str,
		payloadType: proto.RecordTypes_STRING,
	}
}

func NewRecordBuilderFromHex(hex string) RecordBuilder {
	return RecordBuilder{
		payload:     hex,
		payloadType: proto.RecordTypes_HEX,
	}
}

func NewRecordBuilderFromJSON(json string) RecordBuilder {
	return RecordBuilder{
		payload:     json,
		payloadType: proto.RecordTypes_JSON,
	}
}

func NewRecordBuilderFromFile(file_bytes []byte) RecordBuilder {
	return RecordBuilder{
		payload:     file_bytes,
		payloadType: proto.RecordTypes_FILE,
	}
}

func NewRecordBuilderFromBytes(bytes []byte) RecordBuilder {
	return RecordBuilder{
		payload:     bytes,
		payloadType: proto.RecordTypes_BYTES,
	}
}

File: languages/java/src/main/java/com/bloock/sdk/builder/Builder.java
package com.bloock.sdk.builder;

import com.bloock.sdk.bridge.Bridge;
import com.bloock.sdk.bridge.proto.RecordOuterClass;
import com.bloock.sdk.bridge.proto.RecordOuterClass.RecordBuilderFromBytesRequest;
import com.bloock.sdk.bridge.proto.RecordOuterClass.RecordBuilderFromFileRequest;
import com.bloock.sdk.bridge.proto.RecordOuterClass.RecordBuilderFromHexRequest;
import com.bloock.sdk.bridge.proto.RecordOuterClass.RecordBuilderFromJSONRequest;
import com.bloock.sdk.bridge.proto.RecordOuterClass.RecordBuilderFromLoaderRequest;
import com.bloock.sdk.bridge.proto.RecordOuterClass.RecordBuilderFromRecordRequest;
import com.bloock.sdk.bridge.proto.RecordOuterClass.RecordBuilderFromStringRequest;
import com.bloock.sdk.bridge.proto.RecordOuterClass.RecordBuilderResponse;
import com.bloock.sdk.bridge.proto.RecordOuterClass.RecordTypes;
import com.bloock.sdk.bridge.proto.Shared.Error;
import com.bloock.sdk.config.Config;
import com.bloock.sdk.entity.Decrypter;
import com.bloock.sdk.entity.Encrypter;
import com.bloock.sdk.entity.Loader;
import com.bloock.sdk.entity.Record;
import com.bloock.sdk.entity.Signer;
import com.google.protobuf.ByteString;

public class Builder {
  Object payload;
  RecordTypes type;
  RecordOuterClass.Signer signer;
  RecordOuterClass.Encrypter encrypter;
  RecordOuterClass.Decrypter decrypter;

  Builder(Object payload, RecordTypes type) {
    this.payload = payload;
    this.type = type;
  }

  public static Builder fromRecord(Record record) {
    return new Builder(record.toProto(), RecordTypes.RECORD);
  }

  public static Builder fromString(String str) {
    return new Builder(str, RecordTypes.STRING);
  }

  public static Builder fromHex(String str) {
    return new Builder(str, RecordTypes.HEX);
  }

  public static Builder fromJson(String json) {
    return new Builder(json, RecordTypes.JSON);
  }

  public static Builder fromFile(byte[] file) {
    return new Builder(file, RecordTypes.FILE);
  }

  public static Builder fromBytes(byte[] bytes) {
    return new Builder(bytes, RecordTypes.BYTES);
  }

  public static Builder fromLoader(Loader loader) {
    return new Builder(loader.toProto(), RecordTypes.LOADER);
  }

  public Builder withSigner(Signer signer) {
    this.signer = signer.toProto();
    return this;
  }

  public Builder withEncrypter(Encrypter encrypter) {
    this.encrypter = encrypter.toProto();
    return this;
  }

  public Builder withDecrypter(Decrypter decrypter) {
    this.decrypter = decrypter.toProto();
    return this;
  }

  public Record build() throws Exception {
    Bridge bridge = new Bridge();

    RecordBuilderResponse response;

    switch (this.type) {
      case BYTES:
        {
          RecordBuilderFromBytesRequest.Builder builder =
              RecordBuilderFromBytesRequest.newBuilder()
                  .setConfigData(Config.newConfigData())
                  .setPayload(ByteString.copyFrom((byte[]) this.payload));

          if (this.signer != null) {
            builder.setSigner(this.signer);
          }
          if (this.encrypter != null) {
            builder.setEncrypter(this.encrypter);
          }
          if (this.decrypter != null) {
            builder.setDecrypter(this.decrypter);
          }

          response = bridge.getRecord().buildRecordFromBytes(builder.build());
          break;
        }
      case FILE:
        {
          RecordBuilderFromFileRequest.Builder builder =
              RecordBuilderFromFileRequest.newBuilder()
                  .setConfigData(Config.newConfigData())
                  .setPayload(ByteString.copyFrom((byte[]) this.payload));

          if (this.signer != null) {
            builder.setSigner(this.signer);
          }
          if (this.encrypter != null) {
            builder.setEncrypter(this.encrypter);
          }
          if (this.decrypter != null) {
            builder.setDecrypter(this.decrypter);
          }

          response = bridge.getRecord().buildRecordFromFile(builder.build());
          break;
        }
      case HEX:
        {
          RecordBuilderFromHexRequest.Builder builder =
              RecordBuilderFromHexRequest.newBuilder()
                  .setConfigData(Config.newConfigData())
                  .setPayload((String) this.payload);

          if (this.signer != null) {
            builder.setSigner(this.signer);
          }
          if (this.encrypter != null) {
            builder.setEncrypter(this.encrypter);
          }
          if (this.decrypter != null) {
            builder.setDecrypter(this.decrypter);
          }

          response = bridge.getRecord().buildRecordFromHex(builder.build());
          break;
        }
      case JSON:
        {
          RecordBuilderFromJSONRequest.Builder builder =
              RecordBuilderFromJSONRequest.newBuilder()
                  .setConfigData(Config.newConfigData())
                  .setPayload((String) this.payload);

          if (this.signer != null) {
            builder.setSigner(this.signer);
          }
          if (this.encrypter != null) {
            builder.setEncrypter(this.encrypter);
          }
          if (this.decrypter != null) {
            builder.setDecrypter(this.decrypter);
          }

          response = bridge.getRecord().buildRecordFromJson(builder.build());
          break;
        }
      case RECORD:
        {
          RecordBuilderFromRecordRequest.Builder builder =
              RecordBuilderFromRecordRequest.newBuilder()
                  .setConfigData(Config.newConfigData())
                  .setPayload((RecordOuterClass.Record) this.payload);

          if (this.signer != null) {
            builder.setSigner(this.signer);
          }
          if (this.encrypter != null) {
            builder.setEncrypter(this.encrypter);
          }
          if (this.decrypter != null) {
            builder.setDecrypter(this.decrypter);
          }

          response = bridge.getRecord().buildRecordFromRecord(builder.build());
          break;
        }
      case STRING:
        {
          RecordBuilderFromStringRequest.Builder builder =
              RecordBuilderFromStringRequest.newBuilder()
                  .setConfigData(Config.newConfigData())
                  .setPayload((String) this.payload);

          if (this.signer != null) {
            builder.setSigner(this.signer);
          }
          if (this.encrypter != null) {
            builder.setEncrypter(this.encrypter);
          }
          if (this.decrypter != null) {
            builder.setDecrypter(this.decrypter);
          }

          response = bridge.getRecord().buildRecordFromString(builder.build());
          break;
        }
      case LOADER:
        {
          RecordBuilderFromLoaderRequest.Builder builder =
              RecordBuilderFromLoaderRequest.newBuilder()
                  .setConfigData(Config.newConfigData())
                  .setLoader((RecordOuterClass.Loader) this.payload);

          if (this.signer != null) {
            builder.setSigner(this.signer);
          }
          if (this.encrypter != null) {
            builder.setEncrypter(this.encrypter);
          }
          if (this.decrypter != null) {
            builder.setDecrypter(this.decrypter);
          }

          response = bridge.getRecord().buildRecordFromLoader(builder.build());
          break;
        }
      default:
        throw new Exception("Invalid type");
    }

    if (response.getError() != Error.getDefaultInstance()) {
      throw new Exception(response.getError().getMessage());
    }

    return Record.fromProto(response.getRecord());
  }
}
